structs.py

```diff
@@ -0,0 +1,550 @@
+    hostname: str | None
+    dns_name: str | None
+    ip_addresses: list[str] | None
+    instance_id: str | None
+    display_name: str | None
+    machine_status: StatusInfoContents | None
+    modification_status: StatusInfoContents | None
+    base: FormattedBase | None
+    network_interfaces: dict[str, NetworkInterface] | None
+    containers: dict[str, MachineStatus] | None
+    constraints: str | None
+    hardware: str | None
+    controller_member_status: str | None
+    ha_primary: bool | None
+    lxd_profiles: dict[str, LxdProfileContents] | None
+
+    @classmethod
+    def from_dict(cls, d: dict[str, Any]) -> MachineStatus:
+        return cls(
+            juju_status=StatusInfoContents.from_dict(d['juju-status']) if 'juju-status' in d else None,
+            hostname=d.get('hostname'),
+            dns_name=d.get('dns-name'),
+            ip_addresses=[x for x in d['ip-addresses']] if 'ip-addresses' in d else None,
+            instance_id=d.get('instance-id'),
+            display_name=d.get('display-name'),
+            machine_status=StatusInfoContents.from_dict(d['machine-status']) if 'machine-status' in d else None,
+            modification_status=StatusInfoContents.from_dict(d['modification-status']) if 'modification-status' in d else None,
+            base=FormattedBase.from_dict(d['base']) if 'base' in d else None,
+            network_interfaces={k: NetworkInterface.from_dict(v) for k, v in d['network-interfaces'].items()} if 'network-interfaces' in d else None,
+            containers={k: MachineStatus.from_dict(v) for k, v in d['containers'].items()} if 'containers' in d else None,
+            constraints=d.get('constraints'),
+            hardware=d.get('hardware'),
+            controller_member_status=d.get('controller-member-status'),
+            ha_primary=d.get('ha-primary'),
+            lxd_profiles={k: LxdProfileContents.from_dict(v) for k, v in d['lxd-profiles'].items()} if 'lxd-profiles' in d else None,
+        )
+
+
+@dataclasses.dataclass
+class RemoteEndpoint:
+    interface: str
+    role: str
+
+    @classmethod
+    def from_dict(cls, d: dict[str, Any]) -> RemoteEndpoint:
+        return cls(
+            interface=d['interface'],
+            role=d['role'],
+        )
+
+
+@dataclasses.dataclass
+class RemoteAppStatus:
+    url: str
+    endpoints: dict[str, RemoteEndpoint] | None
+    life: str | None
+    app_status: StatusInfoContents | None
+    relations: dict[str, list[str]] | None
+
+    @classmethod
+    def from_dict(cls, d: dict[str, Any]) -> RemoteAppStatus:
+        return cls(
+            url=d['url'],
+            endpoints={k: RemoteEndpoint.from_dict(v) for k, v in d['endpoints'].items()} if 'endpoints' in d else None,
+            life=d.get('life'),
+            app_status=StatusInfoContents.from_dict(d['application-status']) if 'application-status' in d else None,
+            relations={k: [x for x in v] for k, v in d['relations'].items()} if 'relations' in d else None,
```

Comment:
```suggestion
            relations={k: list(v) for k, v in d['relations'].items()} if 'relations' in d else None,
```

if it's needed at all... why is this needed?

Reply:
Good point -- simplistic generator. Fixed. And I think in this context just `v` is fine -- don't even need a copy.